Angular-CLI: Upgrade Angular-CLI version from 5 to 7

sanketbansal opened the pull request:
<!--
(Thanks for sending a pull request! .)
-->

<!--The PR title should start with "Fix #bugnum: " (if applicable), followed by a clear one-line present-tense summary of the changes introduced in the PR. For example: "Fix #bugnum: Introduce the first version of the collection editor." -->

<!-- Add issue number here. If you do not solve the issue entirely, please change the message e.g. "Addresses #IssueNumber -->

#### Changes proposed in this pull request:
- Project has been upgraded to latest angular-cli version.
- Required config files has been updated according to the latest version.
- Angular version related bug present in api.service.ts has been removed.
<!-- Changes: Add here what changes were made in this issue and if possible provide links. -->
<!-- Demo Link: Add here the link where you changes can be seen. -->
- Link to live demo: http://pr-125-evalai.surge.sh <!-- Replace XXX with your PR no: -->

<!-- Screenshots for the change: Add here the screenshot of the fix. -->

sanketbansal committed:
Upgraded Angular-CLI version

Project has been upgraded to latest angular-cli version.
Required config files has been updated according to the latest version.
Angular version related bug present in api.service.ts has been removed.

sanketbansal committed:
Changed typescript version

Typescript version changed to <3.3.0 to meet the requirement of @angular/compiler-cli.

sanketbansal committed:
Fixed e2e testing issue

Downgraded the protractor version to "~5.1.2".

sanketbansal committed:
Fixed production build issue

Downgraded the @angular/compiler-cli version to ^5.2.0.

sanketbansal committed:
Fixed testing errors

Added includePaths property for stylePreprocessorOptions  in angular.json .
Now , ng test command executes successfully.

Shekharrajak commented:
Is this PR completed ? Please ping us after completing the PRs, so that we can review it. I see none of the PRs are merged. Also you can use [WIP] in title of the PR to let us know that PR is not completed yet.

Shekharrajak commented on angular.json:
```diff
@@ -0,0 +1,166 @@
+            "scripts": []
+          },
+          "configurations": {
+            "staging": {
```
We need to document it in development workflow - how to use different configuration/environements.

Shekharrajak commented on package.json:
```diff
@@ -50,6 +52,6 @@
     "ts-node": "~4.1.0",
     "tslint": "^5.9.1",
-    "typescript": "^2.5.3"
+    "typescript": "<3.3.0"
```
Why `<`?

sanketbansal replied:
angular-compiler/cli has some peer dependency on typescript version which requires version < 3.3.0
2.5.3 was failing for latest angular/cli version.

Shekharrajak replied:
I mean why `<`,  isn't latest version of CLI will support latest version of typescript as well? So it can be `>` or `~`

sanketbansal replied:
yeah, CLI supports latest version but some packages has **peer dependency** over **angualr-compiler/cli version** **5.2.0** which itself not supporting typescript versions **>= 3.3.0**

sanketbansal replied:
Typescript version "~3.2.0" should also work.

Shekharrajak reviewed (commented):
Please try to write comments on your changes(if there is any specific reason) so that we can review it easily.

RishabhJain2018 commented on tsconfig.json:
```diff
@@ -14,6 +15,8 @@
+    "baseUrl": "./"
   }
-}
+}
```
Please add a new line here.

sanketbansal replied:
Added new a line

sanketbansal committed:
Added a new line

A new line has been added at the end of tsconfig.json file.

Shekharrajak commented:
I hope you also followed the steps given here https://update.angular.io

If this PR is complete we can merge it but in later PRs make sure you use/update Angular 7 features.

sanketbansal commented:
yeah, I used https://update.angular.io/ for updating angular version.
Okay, I will use Angular 7 features in later PRs.

Shekharrajak commented:
So it can be merged? (repeating the question since I didn't get a clear answer - asked the same thing 1 week [before](https://github.com/Cloud-CV/EvalAI-ngx/pull/125#issuecomment-490762228))

sanketbansal commented:
yes, It can be merged

Shekharrajak commented:
@lunayach @RishabhJain2018 let's merge this. If something went wrong then hopefully @sanketbansal will take care of it in other open PRs.

sanketbansal committed:
Changed typescript version

Typescript version changed from <3.3.0 to ~3.2.0

sanketbansal commented:
I think this PR should not give any issue as such unless another open PR's has some dependency on older versions .

Shekharrajak commented on angular.json:
```diff
@@ -0,0 +1,166 @@
+              "vendorChunk": false,
+              "buildOptimizer": true,
+              "fileReplacements": [
+                {
```
I hope PR https://github.com/Cloud-CV/EvalAI-ngx/pull/127 changes will reflect here

lunayach reviewed (commented):
@sanketbansal Are you sure that everything has been upgraded? In `package.json`, I only see an upgrade in development dependency of `@angular/cli` (i.e. from **1.6.5** to **^7.3.8**). But `@angular/core, animations etc.` still have a version of **5.2**.

Shekharrajak commented:
> @sanketbansal Are you sure that everything has been upgraded? In `package.json`, I only see an upgrade in development dependency of `@angular/cli` (i.e. from **1.6.5** to **^7.3.8**). But `@angular/core, animations etc.` still have a version of **5.2**.

Yes! Using latest version of CLI we can upgrade the project to Angular 7 project, but it would be better to use latest versions of all the dependencies.

This should be done and merged before Sunday.

RishabhJain2018 commented:
@Shekharrajak @lunayach Please feel free to mention me once it is ready.

sanketbansal commented:
Actually I did not updated the angular dependencies because some packages supports older version of those I.e ngx-twitter-timeline package.

Shekharrajak commented:
> Actually I did not updated the angular dependencies because some packages supports older version of those I.e ngx-twitter-timeline package.

Let us know when you will be able to complete it (as I said [here](https://github.com/Cloud-CV/EvalAI-ngx/pull/125#issuecomment-493652445))

sanketbansal commented:
Currently, I am getting a testing issue after updating dependency I will make a commit after resolving it.

sanketbansal committed:
Updated to Angular 7 Version

Updated angular dependencies to version 7 along with 
peer dependencies.

sanketbansal commented:
This is the error i am getting on running **ng test** command.
Any idea how to resolve it?
![Screenshot from 2019-05-24 01-09-07](https://user-images.githubusercontent.com/17106489/58304059-c1324200-7dc0-11e9-8151-fb7ae56c4361.png)

Shekharrajak commented on tsconfig.json:
```diff
@@ -14,6 +15,8 @@
       "dom"
-    ]
+    ],
+    "module": "es2015",
```
What does `module` option do?

Shekharrajak commented on tslint.json:
```diff
@@ -18,7 +18,6 @@
     "import-blacklist": [
       true,
-      "rxjs",
       "rxjs/Rx"
```
`rxjs` will contain `rxjs/Rx` right?

Shekharrajak commented on tsconfig.json:
```diff
@@ -14,6 +15,8 @@
       "dom"
-    ]
+    ],
+    "module": "es2015",
```
What does `module` option do?

sanketbansal replied:
this change is made by ng update command.
I also checked by removing this change but the same error is coming.

Shekharrajak commented on tslint.json:
```diff
@@ -18,7 +18,6 @@
     "import-blacklist": [
       true,
-      "rxjs",
       "rxjs/Rx"
```
`rxjs` will contain `rxjs/Rx` right?

sanketbansal replied:
Ng update made these changes.

Shekharrajak reviewed (commented):
@sanketbansal , for `ng test` error. I  think you need to change `participants`  to ` /participants`  in endpoint service.

sanketbansal committed:
Change to debug testing issue

Changes made in contact-component to resolve testing issue

sanketbansal commented:
> @sanketbansal , for `ng test` error. I think you need to change `participants` to ` /participants` in endpoint service.

I tried but that is not working.
Also, I do not see any routing being done to "participate" by contact-component
I fixed the 'auth/login' problem but it is giving the same error again.

sanketbansal committed:
Made rxjs changes

Added some  rxjs changes to resolve testing issue

Shekharrajak commented:
> > @sanketbansal , for `ng test` error. I think you need to change `participants` to ` /participants` in endpoint service.
> 
> I tried but that is not working.
> Also, I do not see any routing being done to "participate" by contact-component
> I fixed the 'auth/login' problem but it is giving the same error again.

What are the other relevant ways did you find in google? 

Can you try this in routing module:

```

const routes_child: Routes = [
  {
    path: 'auth',
    component: AuthComponent,
    children: [
      {path: '', redirectTo: 'login', pathMatch: 'full'},
      {path: 'login', component: LoginComponent},
      {path: 'signup', component: SignupComponent},
      {path: 'verify-email/:token', component: VerifyEmailComponent},
      {path: '**', redirectTo: 'login'}
    ]
  },
  {
    path: 'challenge/:id',
    component: ChallengeComponent,
    children: [
      {path: '', redirectTo: 'overview', pathMatch: 'full'},
      {path: 'overview', component: ChallengeoverviewComponent},
      {path: 'evaluation', component: ChallengeevaluationComponent},
      {path: 'phases', component: ChallengephasesComponent},
      {path: 'participate', component: ChallengeparticipateComponent},
      {path: 'submit', component: ChallengesubmitComponent},
      {path: 'submissions', component: ChallengesubmissionsComponent},
      {path: 'submissions/:phase', component: ChallengesubmissionsComponent},
      {path: 'submissions/:phase/:submission', component: ChallengesubmissionsComponent},
      {path: 'leaderboard', component: ChallengeleaderboardComponent},
      {path: 'leaderboard/:split', component: ChallengeleaderboardComponent},
      {path: 'leaderboard/:split/:entry', component: ChallengeleaderboardComponent}
    ]
  },
  {
    path: 'challenges',
    component: PubliclistsComponent,
    children: [
      {path: '', redirectTo: 'all', pathMatch: 'full'},
      {path: 'all', component: ChallengelistComponent},
      {path: 'me', component: ChallengelistComponent}
    ]
  },
    {
    path: 'teams',
    component: PubliclistsComponent,
    children: [
      {path: '', redirectTo: 'participants', pathMatch: 'full'},
      {path: 'participants', component: TeamlistComponent},
      {path: 'hosts', component: TeamlistComponent}
    ]
  },
];
```

and import it into NgModule:

```
imports: [ RouterModule.forRoot(routes), RouterModule.forcChild(routes_child) ],
```

Also you can put `/` in EndpointsService at starting of path variables: 

```
  participants = '/participants/';
  hosts = '/hosts/';
  jobs = '/jobs/';
  challenges = '/challenges/';
  challenge = '/challenge/';
  auth = '/auth/';
  analytics = '/analytics/';

```
Is it `ng test` running again once it complete ?

sanketbansal committed:
Added routing in challenge create test case file

Testing routes are added for challenge-create component to resolve teams/hosts error .

sanketbansal committed:
Added routing in test case files

Routes are added in the test suites to resolve 'cannot find teams/hosts' issue

sanketbansal committed:
Added routing in testing files to resolve testing issue

Routes are added in the test suite to resolve 'cannot find url participate' issue

sanketbansal committed:
Added routing to resolve testing issue

Routes are added in the test suite to resolve 'cannot find url auth/login' issue

sanketbansal committed:
Fixed Linting issue

Added trailing white space in app.component.ts file

sanketbansal commented:
**Fixed** testing issue please review

lunayach reviewed (approved):
Great job @sanketbansal pulling it through. Let’s merge this finally 🎉 @Shekharrajak @RishabhJain2018

lunayach committed:
Merge branch 'master' into upgrade

Shekharrajak commented on src/app/components/contact/contact.component.spec.ts:
```diff
@@ -9,46 +9,66 @@ import { GlobalService } from '../../services/global.service';
 
   it('should create', () => {
-    expect(component).toBeTruthy();
+    fixture.ngZone.run(() => {
```
Bit confused about these lines. Can you please @sanketbansal write about this change?

sanketbansal replied:
Angular gives a warning if you perform routing **outside angular zone** so called the **ngZone.run** method to perform routing inside it.

Shekharrajak commented on src/app/components/publiclists/challengelist/challengelist.component.spec.ts:
```diff
@@ -1,43 +1,77 @@
+    router = TestBed.get(Router);
     fixture = TestBed.createComponent(ChallengelistComponent);
+    authService = TestBed.get(AuthService);
+    globalService = TestBed.get(GlobalService);
```
I hope you will use `globalService` somewhere in later PRs.

sanketbansal replied:
yeah, It will be used later at the time of adding more test cases.

Shekharrajak reviewed (approved):
Good work! Lets merge it, so that you can work on other PRs.

RishabhJain2018 commented on src/app/components/challenge/challengesubmit/challengesubmit.component.spec.ts:
```diff
@@ -7,31 +7,56 @@ import { GlobalService } from '../../../services/global.service';
+    fixture.ngZone.run(() => {
+      router.navigate(['/challenge/0/submit']).then(() => {
+        fixture.detectChanges();
+        console.log(router.url);
```
Please remove this console statement

sanketbansal replied:
Done

sanketbansal committed:
Removed logs 

Removed console log lines from challenge submit component

RishabhJain2018 committed:
Merge branch 'master' into upgrade

Shekharrajak commented:
Angular CLI version 8 is released (10 hours ago): https://www.npmjs.com/package/@angular/cli/v/8.0.0, I have restarted the travis build. if you find any error then please have a look @sanketbansal

RishabhJain2018 committed:
Merge branch 'master' into upgrade
